File: src/util/random.ts
import { create } from 'random-seed'
import config from '../config'

const baseSeedKey = config.discord.botToken as string

export function randomInclusive(min: number, max: number, seed?: string): number {
  if (!seed) {
    // make random inclusive
    return Math.floor(Math.random() * (max - min + 1)) + min
  }

  return create(seed + baseSeedKey).intBetween(min, max)
}

/*

Taken from https://stackoverflow.com/a/46774731
After a 100 iterations you should see roughly
    ==================
    | Result | Times |
    ==================
    |      1 |    55 |
    |      2 |    28 |
    |      3 |     8 |
    |      4 |     7 |
    |      5 |     2 |
    ==================

    note: 49 + weightedRandom(1, 51) if you want to get numbers between 50 and 100
*/
export function weightedRandom(min: number, max: number, seed?: string): number {
  const rand = seed ? create(seed + baseSeedKey).random() : Math.random()
  return Math.round(max / (rand * max + min))
}

File: src/util/random.rs
use rand::prelude::*;
use rand_chacha::ChaCha8Rng;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// Generate a random inclusive integer between min and max (unseeded).
pub fn random_inclusive(min: i64, max: i64) -> i64 {
    let mut rng = rand::thread_rng();
    rng.gen_range(min..=max)
}

/// Generate a random inclusive integer between min and max using a seeded RNG.
/// The seed string is combined with `base_seed` to produce deterministic results.
pub fn seeded_random_inclusive(min: i64, max: i64, seed: &str, base_seed: &str) -> i64 {
    let mut rng = make_seeded_rng(seed, base_seed);
    rng.gen_range(min..=max)
}

/// Generate a weighted random number between min and max using a seeded RNG.
pub fn seeded_weighted_random(min: i64, max: i64, seed: &str, base_seed: &str) -> i64 {
    let mut rng = make_seeded_rng(seed, base_seed);
    let rand: f64 = rng.r#gen();
    (max as f64 / (rand * max as f64 + min as f64)).round() as i64
}

fn make_seeded_rng(seed: &str, base_seed: &str) -> ChaCha8Rng {
    let combined = format!("{}{}", seed, base_seed);
    let mut hasher = DefaultHasher::new();
    combined.hash(&mut hasher);
    let hash = hasher.finish();
    ChaCha8Rng::seed_from_u64(hash)
}
